README.md
# XCListen
[![Build Status](https://travis-ci.org/mneorr/xclisten.png?branch=master)](https://travis-ci.org/mneorr/xclisten)
[![Code Climate](https://codeclimate.com/github/mneorr/xclisten.png)](https://codeclimate.com/github/mneorr/xclisten)

A __zero-configuration__ file watcher for ObjectiveC.


It:
- runs tests each time you save a file
- runs `pod install` each time you save the Podfile
- formats your tests with RSpec-style output using [XCPretty](https://github.com/mneorr/xcpretty)


## Installation

```
$ gem install xclisten
```

## Usage

Run this command from the root of your repository:
```
$ xclisten
```
Simple, huh?

![](http://i.imgur.com/JpsMMBW.gif)

If you have an OSX project, you'll want to run it with `--osx` flag.

## Flags

``` bash
Usage: xclisten [optional flags]
        --osx                        Run with OSX sdk (without simulator)
        --ios                        [DEFAULT] Run with iOS sdk
    -d, --device                     Simulated device [iphone5s, iphone5, iphone4]. Default is iphone5s
    -s, --scheme SCHEME              BYOS (Bring your own scheme)
    -w, --workspace WORKSPACE        BYOW (Bring your own workspace)
    -h, --help                       Show this message
    -v, --version                    Show version
```

## Something went wrong!

No worries, just `tail -f xcodebuild_error.log` and let us know what's happening.

## Acknowledgements

This project is inspired by many general-purpose listeners out there,
such as [kicker](https://github.com/alloy/kicker) and [Guard](https://github.com/guard/guard).


## TODO

- Support non-cocoapods projects, when there's no .xcworkspace
- Acceptance tests


CHANGELOG.md
# Changelog

# 0.0.8

##### Bug Fixes

* Fixed building for OSX. We were specifying `destination` which was bad. |
  [#17](https://github.com/mneorr/xclisten/issues/17)
* Fixed Shellwords import that was crashing system ruby (2.0.0) |
  [#15](https://github.com/mneorr/xclisten/issues/15)
* Fixed devices flags |
  [#16](https://github.com/mneorr/xclisten/issues/16)


# 0.0.7

##### Enhancements

* Forgot to escape schemes & workspaces in 0.0.6. done.


# 0.0.6

##### Enhancements

* Added iPads (`xclisten -d [ipad2, ipad4, ipad_air]`)

# 0.0.5

##### Bug Fixes

* Fixed workspace and device flags


# 0.0.4

##### Enhancements

* Search for `.xcworkspace` recursively |
  [Delisa Mason](https://github.com/kattrali) |
  [#8](https://github.com/mneorr/xclisten/pull/8)
* Added some tests and CI integration \o/
* Added `--workspace -w` flag to specify a custom workspace location
* Specified `macosx` sdk rather than nothing on OSX builds


# 0.0.3

##### Enhancements

* Use lower-level fork and exec for spawning xcodebuild
* Support for custom schemes
* Support for custom devices (iphone5s, iphone5, iphone4)

# 0.0.2

##### Bug Fixes

* Kill child processes on Ctrl-C


# 0.0.1

* Initial release



spec/lib/xclisten_spec.rb
require 'xclisten'

class XCListen

  describe XCListen do

    class ShellTask
      def run(args)
        @command = args
      end
    end

    before(:all) do
      @base_dir = Dir.pwd
    end

    after(:each) do
      Dir.chdir(@base_dir)
    end

    context "Devices" do
      let(:devices) { XCListen::IOS_DEVICES }

      it "has ALL of them" do
        devices['iphone5s'].should == 'iPhone Retina (4-inch 64-bit)'
        devices['iphone5'].should == 'iPhone Retina (4-inch)'
        devices['iphone4'].should == 'iPhone Retina (3.5-inch)'
        devices['ipad2'].should == 'iPad'
        devices['ipad4'].should == 'iPad Retina'
        devices['ipad_air'].should == 'iPad Retina (64-bit)'
      end
    end

    context "Defaults" do

      before(:each) do
        Dir.chdir('fixtures/ObjectiveRecord')
        @xclisten = XCListen.new
      end

      it "runs runs tests on iPhone 5s by default" do
        @xclisten.device.should == 'iPhone Retina (4-inch 64-bit)'
      end

      it "uses the project name as the default scheme" do
        @xclisten.scheme.should == 'SampleProject'
      end

      it "uses the first found workspace by default" do
        @xclisten.workspace.should == 'Example/SampleProject.xcworkspace'
      end

      it "uses iphonesimulator sdk by default" do
        @xclisten.sdk.should == 'iphonesimulator'
      end

      it "listens to the closest workspace instead of alphabetically" do
        Dir.chdir("#{@base_dir}/fixtures/MadeUpProject")
        XCListen.new.workspace.should == 'MadeUpProject.xcworkspace'
      end

    end

    it "initializes with device" do
      xclisten = XCListen.new(:device => 'iphone5')
      xclisten.device.should == 'iPhone Retina (4-inch)'
    end

    it "initializes with a different workspace" do
      xclisten = XCListen.new(:workspace => 'Example/Sample.xcworkspace')
      xclisten.workspace.should == 'Example/Sample.xcworkspace'
    end

    it "initializes with a different scheme" do
      xclisten = XCListen.new(:scheme => 'AnotherScheme')
      xclisten.scheme.should == 'AnotherScheme'
    end

    it 'initializes with SDK' do
      xclisten = XCListen.new(:sdk => 'iphonesimulator')
      xclisten.sdk.should == 'iphonesimulator'
    end

    it 'uses xcodebuild with given flags' do
      flags = {
        :workspace => 'Example/Sample.xcworkspace',
        :scheme => 'kif',
        :sdk => 'iphoneos',
        :device => 'name=iPhone Retina (4-inch 64-bit)'
      }
      xclisten = XCListen.new(flags)
      xclisten.xcodebuild.should == "xcodebuild -workspace #{flags[:workspace]} -scheme #{flags[:scheme]} -sdk #{flags[:sdk]} -destination '#{flags[:device]}'"
    end

    it "doesn't specify destination for OSX" do
      xclisten = XCListen.new(:sdk => 'macosx')
      xclisten.xcodebuild.should_not include('-destination')
    end

    it "knows if it can run" do
      Dir.stub(:glob).and_return([""])
      xclisten = XCListen.new
      xclisten.can_run?.should be_false
    end

    context "finding the right workspace" do

      it "works with one level deep workspace" do
        Dir.chdir('fixtures/ObjectiveRecord')
        XCListen.new.workspace.should == 'Example/SampleProject.xcworkspace'
      end

      it "works with xcworkspace in the same directory" do
        Dir.chdir('fixtures/AFNetworking')
        XCListen.new.workspace.should == 'AFNetworking.xcworkspace'
      end
    end
  end

end



features/steps/xclisten_steps.rb
Given(/^I am in a directory with an xcworkspace$/) do
  set_run_path('dir_with_xcworkspace')
end

Given(/^I am in a directory with a nested xcworkspace$/) do
  set_run_path('dir_with_nested_xcworkspace')
end

Given(/^I am in a directory with multiple xcworkspaces$/) do
  set_run_path('dir_with_multiple_xcworkspaces')
end

Given(/^I am in a directory without an xcworkspace$/) do
  set_run_path('dir_without_xcworkspace')
end

When(/^I run xclisten with flags "(.*?)"$/) do |flags|
  run_xclisten(flags)
end

Then(/^xclisten should be listening to the xcworkspace in the current directory$/) do
  xcworkspace_path.should =~ /Sample.xcworkspace$/
end

Then(/^xclisten should be listening to the xcworkspace nested in the current directory$/) do
  xcworkspace_path.should =~ /NestedSample.xcworkspace$/
end

Then(/^xclisten should be listening to the first xcworkspace listed alphabetically$/) do
  xcworkspace_path.should =~ /A.xcworkspace$/
end

Then(/^I should get an error message about no xcworkspace being found$/) do
  run_output.should =~ /^\[!\] No xcworkspace found in this directory/
end

lib/xclisten.rb
require 'xclisten/version'
require 'xclisten/shell_task'
require 'listen'

class XCListen

  attr_reader :device
  attr_reader :scheme
  attr_reader :sdk
  attr_reader :workspace

  def initialize(opts = {})
    @scheme = opts[:scheme] || project_name
    @sdk = opts[:sdk] || 'iphonesimulator'
    @workspace = opts[:workspace] || workspace_path
    @device = IOS_DEVICES[opts[:device]] || IOS_DEVICES['iphone5s']
  end

  IOS_DEVICES = {
    'iphone5s' => 'iPhone Retina (4-inch 64-bit)',
    'iphone5' => 'iPhone Retina (4-inch)',
    'iphone4' => 'iPhone Retina (3.5-inch)',
    'ipad2' => 'iPad',
    'ipad4' => 'iPad Retina',
    'ipad_air' => 'iPad Retina (64-bit)'
  }

  def workspace_path
    @workspace_path ||= Dir.glob("**/*.xcworkspace").sort_by(&:length).first
  end

  def project_name
    @project_name ||= File.basename(workspace_path, ".*") if workspace_path
  end

  def xcodebuild
    cmd = "xcodebuild -workspace #{workspace} -scheme #{scheme} -sdk #{sdk}"
    cmd += " -destination 'name=#{device}'" unless @sdk == 'macosx'
    cmd
  end

  def install_pods
    Dir.chdir(File.dirname(workspace)) do
      system 'pod install'
      puts 'Giving Xcode some time to index...'
      sleep 10
    end
  end

  def run_tests
    ShellTask.run("#{xcodebuild} test 2> xcodebuild_error.log | xcpretty -tc")
  end

  #TODO TEST THIS SPIKE
  def validate_run
    unless can_run?
      puts "[!] No xcworkspace found in this directory (or any child directories)"
      exit 1
    end
  end

  def can_run?
    project_name && !project_name.empty?
  end

  def listen
    validate_run
    puts "Started xclistening to #{workspace}"
    ignored = [/.git/, /.xc(odeproj|workspace|userdata|scheme|config)/, /.lock$/, /\.txt$/, /\.log$/]

    listener = Listen.to(Dir.pwd, :ignore => ignored) do |modified, added, removed|
      system 'clear'
      if modified.first =~ /Podfile$/
        install_pods
      else
        run_tests
      end
    end

    listener.start
    sleep
  end

end

